go/kbfs/simplefs/downloads.go

```diff
@@ -0,0 +1,353 @@
+	err = m.k.SimpleFSCopyRecursive(ctx, keybase1.SimpleFSCopyRecursiveArg{
+		OpID: opid,
+		Src:  keybase1.NewPathWithKbfs(arg.Path),
+		Dest: keybase1.NewPathWithLocal(downloadPath),
+	})
+	if err != nil {
+		return "", err
+	}
+
+	func() {
+		defer m.publisher.DownloadStatusChanged()
+		m.lock.Lock()
+		defer m.lock.Unlock()
+		m.downloads[downloadID] = download{
+			info: keybase1.DownloadInfo{
+				DownloadID:        downloadID,
+				Path:              arg.Path,
+				Filename:          filename,
+				StartTime:         keybase1.ToTime(time.Now()),
+				IsRegularDownload: arg.IsRegularDownload,
+			},
+			opid: opid,
+			state: keybase1.DownloadState{
+				DownloadID: downloadID,
+			},
+		}
+	}()
+
+	bgCtx, cancelBtCtx := m.makeContext()
+	done := func(err error) {
+		_ = m.updateDownload(downloadID, func(d download) download {
+			if d.state.Done || d.state.Canceled || len(d.state.Error) > 0 {
+				return d
+			}
+			if err == context.Canceled {
+				d.state.Canceled = true
+			} else if err != nil {
+				d.state.Error = err.Error()
+			} else {
+				d.state.EndEstimate = keybase1.ToTime(time.Now())
+				d.state.Progress = 1
+				d.state.Done = true
+			}
+			return d
+		})
+		cancelBtCtx()
+	}
+	go m.monitorDownload(bgCtx, opid, downloadID, done)
+	go m.waitForDownload(bgCtx, downloadID, downloadPath, done)
+
+	return downloadID, nil
+}
+
+func (m *downloadManager) getDownloadStatus(ctx context.Context) (
+	status keybase1.DownloadStatus) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	for _, download := range m.downloads {
+		status.States = append(status.States, download.state)
+		if download.info.IsRegularDownload {
+			status.RegularDownloadIDs = append(
+				status.RegularDownloadIDs, download.info.DownloadID)
+		}
+	}
+	sort.Slice(status.RegularDownloadIDs, func(i, j int) bool {
+		d1, ok := m.downloads[status.RegularDownloadIDs[i]]
+		if !ok {
+			return false
+		}
+		d2, ok := m.downloads[status.RegularDownloadIDs[j]]
+		if !ok {
+			return false
+		}
+		return d1.info.StartTime.After(d2.info.StartTime)
+	})
+	return status
+}
+
+func (m *downloadManager) cancelDownload(
+	ctx context.Context, downloadID string) error {
+	d, err := m.getDownload(downloadID)
+	if err != nil {
+		return err
+	}
+	return m.k.SimpleFSCancel(ctx, d.opid)
+}
+
+func (m *downloadManager) dismissDownload(
+	ctx context.Context, downloadID string) {
+	m.cancelDownload(ctx, downloadID)
+	defer m.publisher.DownloadStatusChanged()
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	delete(m.downloads, downloadID)
+}
+
+func (m *downloadManager) getDownloadInfo(downloadID string) (keybase1.DownloadInfo, error) {
+	d, err := m.getDownload(downloadID)
+	if err != nil {
+		return keybase1.DownloadInfo{}, err
+	}
+	return d.info, nil
+}
+
+func (m *downloadManager) configureDownload(cacheDirOverride string, downloadDirOverride string) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if len(cacheDirOverride) > 0 {
+		m.cacheDir = cacheDirOverride
```

Comment:
I wonder if we should try `mkdir`ing and/or `stat`ing these new directories, to make sure they are valid (would require returning an error from this function).  Not sure what the right semantics are.

Reply:
Actually we already make the dir in getDownloadPath(), so we should be fine.

Reply:
Got it! I think we're fine here; if there's an error writing into that dir we'd eventually get an error for the download, which is probably what we want anyway as this way we wouldn't bother user with download related problems until they want to download something.

Reply:
Ok cool.  Though if we ever let the user set their own configurable directory, we might want to reconsider that.  Ok for now though!